Topdesk mirror integration

atossz opened the pull request:
<!-- REMINDER: THIS IS A PUBLIC REPO DO NOT POST HERE SECRETS/SENSITIVE DATA -->

## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
NA

## Description
We have build the same mirror integration for TOPdesk as was implemented within ServiceNow.
We need this to get a generic way to communicate with systems like TOPdesk and ServiceNow.
Several fields, comments, notes are being mirrored between Xsoar and TOPdesk.
Mirror integration was planned for 2022Q2, but was postponed to 2022Q3 and we had to continu.

I proposed to add this into a pull request. I need some last help with the testing possibilities of the mirror.
It is working in our environment, but I don't have any experience with the automated tests.

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Cortex XSOAR
- [ ] 6.0.0
- [ ] 6.1.0
- [x] 6.2.0
- [ ] 6.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [X] No

## Must have
- [X] Tests
- [X] Documentation

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/demisto/content?pullRequest=18735) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/demisto/content?pullRequest=18735) before we can accept your contribution.<br/><hr/>**Sander Zumbrink** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/demisto/content?pullRequest=18735) it.</sub>

bakatzir commented:
relates to: https://github.com/demisto/etc/issues/35309

BEAdi commented on Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py:
```diff
@@ -1272,6 +1295,226 @@ def fetch_incidents(client: Client,
+                    xargs['action_invisible_for_caller'] = False
+                # Sometimes user is an empty str, not None, therefore nothing is displayed
+                user = entry.get('user', 'dbot')
+                if (user):
```
Why with parentheses?

atossz replied:
changed

BEAdi commented on Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py:
```diff
@@ -1272,6 +1295,226 @@ def fetch_incidents(client: Client,
+                else:
+                    name = 'Xsoar dbot'
+
+                text = f"<i><u>Update from {name}:</u></i><br><br>{str(entry.get('contents', ''))}" \
+                    + "<br><br><i>Mirrored from Cortex XSOAR</i>"
+                # client.add_comment(ticket_id, ticket_type, key, text)
```
Does the text need to be sent this way?
Why is the `add_comment` call commented?

atossz replied:
I removed the test rule (client.add_comment)

The text is formatted this way, so the user in TOPdesk could see who did the update in Xsoar and the message has been mirrored from Xsoar (requirement from our team).

Is it OK to leave it in this format?

BEAdi replied:
Yes

BEAdi commented on Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py:
```diff
@@ -1272,6 +1295,226 @@ def fetch_incidents(client: Client,
+                    + "<br><br><i>Mirrored from Cortex XSOAR</i>"
+                # client.add_comment(ticket_id, ticket_type, key, text)
+
+                xargs['id'] = ticket_id
```
This can be set when declaring `xargs`

atossz replied:
moved the to declaring xargs

BEAdi commented on Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py:
```diff
@@ -1272,6 +1295,226 @@ def fetch_incidents(client: Client,
+                tags = entry.get('tags', [])
+                if params.get('work_notes_tag') in tags:
+                    xargs['action_invisible_for_caller'] = True
+                elif params.get('comments_tag') in tags:
+                    xargs['action_invisible_for_caller'] = False
```
This can be removed as the default value for `action_invisible_for_caller` is False.

BEAdi commented on Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py:
```diff
@@ -1272,6 +1295,226 @@ def fetch_incidents(client: Client,
+
+    """
+
+    incident_type_scheme = SchemeTypeMapping(type_name=client.ticket_type)
+    demisto.debug(f'Collecting incident mapping for incident type - "{client.ticket_type}"')
```
I see that the only `ticket_type` is `'incident'`, so it can be changed here. The `type_name` here needs to be the name that is in the `yml` of the incident type, so it can't be just `incident`.
```suggestion
    incident_type_scheme = SchemeTypeMapping(type_name='TOPdesk Incident')
    demisto.debug(f'Collecting incident mapping for incident type - TOPdesk Incident')
```

atossz replied:
I copied this from the ServiceNowV2 integration.
For now we only support the incident module in TOPdesk, but maybe later the change.
But that module is much more complex to support.

Is it ok to let is like it was?

BEAdi replied:
No, as it has a different behavior than wanted. The changes is only in XSOAR- so the mapping scheme will be related to the right incident type.

BEAdi reviewed (changes requested):
Hi Sander,
Thank you very much for your contribution! 
Please note this changes:
1. Mirroring integration must hold an outgoing mapper and a classifier, please add them and update the incoming mapper.
2. Add to the integration's README an explanation about the mirroring functionality that you added. You can find an example for such a description [here](https://xsoar.pan.dev/docs/reference/integrations/crowdstrike-falcon#incident-mirroring).
3. Please wrap the `get_remote_data_command` and the `update_remote_system_command` with `try` and `except`, where when mirroring in the error is written to `mirrored_data['in_mirror_error']`, and when mirroring out the error is written to `demisto.error` and the function returns the remote incident id as if the mirroring worked.

atossz commented on Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.yml:
```diff
@@ -5850,12 +5822,20 @@ script:
       type: string
-  dockerimage: demisto/python3:3.9.9.25564
+  - name: get-mapping-fields
+    description: Returns the list of fields for an incident type.
```
changed

atossz commented on Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py:
```diff
@@ -1382,6 +1625,14 @@ def main() -> None:
                                                     demisto_params=demisto_params)
             demisto.setLastRun(last_fetch)
             demisto.incidents(incidents)
+        elif demisto.command() == 'get-remote-data':
+            return_results(get_remote_data_command(client, demisto.args(), demisto.params()))
+        elif demisto.command() == 'update-remote-system':
+            return_results(update_remote_system_command(client, demisto.args(), demisto.params()))
+        elif demisto.command() == 'get-mapping-fields':
+            return_results(get_mapping_fields_command(client))
+        elif demisto.command() == 'get-modified-remote-data':
+            return_results(get_modified_remote_data_command(client, demisto.args(), demisto.params()))
```
changed

atossz commented on Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py:
```diff
@@ -1255,10 +1271,17 @@ def fetch_incidents(client: Client,
 
             incident = {
-                'name': f"TOPdesk incident {topdesk_incident['number']}",
+                'name': f"{topdesk_incident['briefDescription']}",
```
The briefDescription in TOPdesk is sort of the subject of the incident. It is nicer to see that subject than the TOPdesk number. It doesn't say anything if you look through the list of incidents.

Is it ok to keep it as briefDescription?

BEAdi replied:
In TOPdesk, is it shown as a subject? I'm asking to better understand it is will be clear to users that two incidents are related.

atossz replied:
Yes, it is as a subject in TOPdesk.

atossz commented on Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py:
```diff
@@ -23,7 +23,14 @@
 DATE_FORMAT_FULL = "%Y-%m-%dT%H:%M:%S.%f%z"
 MAX_API_PAGE_SIZE = 10000
-FIRST_REST_API_VERSION_WITH_NEW_QUERY = "3.4.0"
+FIRST_REST_API_VERSION_WITH_NEW_QUERY = "3.3.0"
```
I talked with the developers of TOPdesk. They told me the API query functionality is already in version 3.3.0, but it is not documented. They told me the old API documentation was not easy to maintain.

I had some issues in getting the updates for the mirroring and needed in my version also the older API version. So it works with version 3.3.0, but is not in the documentation, but I got conformation it works.

atossz commented on Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py:
```diff
@@ -1255,10 +1271,17 @@ def fetch_incidents(client: Client,
             incident_created_time = last_fetch_datetime
         assert incident_created_time is not None
+        topdesk_incident['mirror_direction'] = MIRROR_DIRECTION.get(str(demisto_params.get('mirror_direction')))
+        topdesk_incident['mirror_tags'] = [
+            demisto_params.get('comment_tag'),
+            demisto_params.get('file_tag'),
+            demisto_params.get('work_notes_tag')
+        ]
```
Copied this code from the ServiceNowV2 integration. Works the same.

atossz replied:
added default values, so None could not happen.

atossz commented on Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py:
```diff
@@ -1272,6 +1295,226 @@ def fetch_incidents(client: Client,
+
+    if not result:
+        demisto.debug('Ticket was not found!')
+        return 'Ticket was not found'
```
Changed

atossz commented on Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py:
```diff
@@ -1272,6 +1295,226 @@ def fetch_incidents(client: Client,
+    ticket_last_update = dateparser.parse(str(ticket["modificationDate"]), settings={'TIMEZONE': 'UTC'})  # type: ignore
+    assert ticket_last_update is not None
+
+    if last_update > ticket_last_update:
+        demisto.debug('Nothing new in the ticket')
+        ticket = {}
+    else:
+        demisto.debug('ticket is updated')
```
Copy from ServiceNowV2 integration

atossz replied:
Changed

atossz commented on Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py:
```diff
@@ -1272,6 +1295,226 @@ def fetch_incidents(client: Client,
+                else:
+                    name = "Unknown"
+
+                date_time = entry_date.strftime("%d-%m-%Y %H:%M:%S")
```
changed

atossz commented on Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py:
```diff
@@ -1272,6 +1295,226 @@ def fetch_incidents(client: Client,
+def get_modified_remote_data_command(client: Client, args: Dict[str, Any], params: Dict) -> GetModifiedRemoteDataResponse:
+    remote_args = GetModifiedRemoteDataArgs(args)
+    query_date = dateparser.parse(remote_args.last_update,
+                                  settings={'TIMEZONE': 'UTC'}).strftime('%Y-%m-%dT%H:%M:%SZ')  # type: ignore
```
changed

atossz commented on Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py:
```diff
@@ -1272,6 +1295,226 @@ def fetch_incidents(client: Client,
+        for key in parsed_args.delta:
+            update_args[key] = parsed_args.delta[key]
+
+        demisto.debug(f'SZU update_args=[{update_args}]')
```
SZU... my short abbreviation... Sander Zumbrink
removed the debug line...

atossz commented on Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py:
```diff
@@ -1272,6 +1295,226 @@ def fetch_incidents(client: Client,
+        update_args = {
+            'id': ticket_id
+        }
+        for key in parsed_args.delta:
+            update_args[key] = parsed_args.delta[key]
```
changed

atossz commented on Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py:
```diff
@@ -1272,6 +1295,226 @@ def fetch_incidents(client: Client,
+            })
+
+    demisto.debug(f'Pull result is {ticket}')
+    return [ticket] + entries
```
Changed

BEAdi reviewed (changes requested):
Thank you for the changes you did!
For the next time, please don't resolve my comments by yourself- it helps me to follow the changes when I am the one resolving them.

Please mind the previous requested changes you still haven't referred, and also add all the mirroring commands you introduced, with their arguments, to the `yml` file.

atossz commented on Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py:
```diff
@@ -1272,6 +1295,226 @@ def fetch_incidents(client: Client,
+    if parsed_args.incident_changed:
+        demisto.debug(f'Incident changed: {parsed_args.incident_changed}')
+        # Close ticket if needed
+        if parsed_args.inc_status == IncidentStatus.DONE and params.get('close_ticket'):
+            # Set status TOPdesk ticket to Closed
+            demisto.debug('Close TOPdesk ticket')
+            # Close with API call or set field and let mirroring handle it.
+            # client.update_incident
```
I'm setting the 'processingStatus' on the ticket to closed.
TOPdesk will close the setting after the sync.

atossz commented:
Sorry... I won't change the status to resolved anymore...
Everything I put on resolved, I changed back again.
I'll commit last change and you can check them.

The last part I have to do is the documentation, mapper files and classifier.

BEAdi reviewed (changes requested):
You've made some great changes!

Summing up what should be done next:
1. Wrap the `update_remote_system_command` with `try` and `except`, where the error is written to `demisto.error` and the function returns the remote incident id as if the mirroring worked. You can see an example for it [here](https://github.com/demisto/content/blob/master/Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py#L1844).
2. Add an outgoing mapper and a classifier, and update the incoming mapper.
3. Add all the mirroring commands you introduced, with their arguments, to the `yml` file.
4. Add to the integration's README an explanation about the mirroring functionality that you added. You can find an example for such a description [here](https://xsoar.pan.dev/docs/reference/integrations/crowdstrike-falcon#incident-mirroring).

atossz commented on Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py:
```diff
@@ -192,7 +206,7 @@ def update_incident(self, args: Dict[str, Any]) -> Dict[str, Any]:
 
         request_params = prepare_touch_request_params(args)
-
+        demisto.debug(f"SZU args=[{args}] json=[{request_params}]")
```
fixed

atossz commented:
@BEAdi ,

I've uploaded the requested files and changes in the code.
I'll continu with the .yml file and the README file.

BEAdi commented on Packs/TOPdesk/Integrations/TOPdesk/README.md:
```diff
@@ -51,7 +51,32 @@ Make sure you use the right account for your needs and that the account used has
+    - Outgoing - Any changes in XSOAR incidents will be reflected in TOPdesk incidents (`tags`, `status`).
+    - Incoming And Outgoing - Changes in XSOAR incidents and TOPdesk incidents will be reflected in both directions.
+    - None - Turns off incident mirroring.
+6. In the *Comment Entry Tag* integration parameter, you can set the tag which can be added to comments to mirror from XSOAR to TOPdesk. The comment in TOPdesk will be visible for the person and the operator. 
+7. In the *Work Note Entry Tag* integration parameter, you can set the tag which can be added to comments to mirror from XSOAR to TOPdesk. The work note in TOPdesk will only be visible for the operator and not the person. 
+8. In the *File Entry Tag* integration parameter, you can set the tag which can be added to a file to mirror the file from XSOAR to TOPdesk. The file in TOPdesk will only be visible for the operator and not the person. 
```
I will be happy to see this in the demo. 6 to 8 can be merged into one instruction.

atossz replied:
merged into 1 instruction

atossz replied:
done

BEAdi commented on Packs/TOPdesk/Integrations/TOPdesk/README.md:
```diff
@@ -51,7 +51,32 @@ Make sure you use the right account for your needs and that the account used has
+6. In the *Comment Entry Tag* integration parameter, you can set the tag which can be added to comments to mirror from XSOAR to TOPdesk. The comment in TOPdesk will be visible for the person and the operator. 
+7. In the *Work Note Entry Tag* integration parameter, you can set the tag which can be added to comments to mirror from XSOAR to TOPdesk. The work note in TOPdesk will only be visible for the operator and not the person. 
+8. In the *File Entry Tag* integration parameter, you can set the tag which can be added to a file to mirror the file from XSOAR to TOPdesk. The file in TOPdesk will only be visible for the operator and not the person. 
+9. Optional: Check the *Close Mirrored XSOAR Incident* integration parameter to close the Cortex XSOAR incident when the corresponding incident is closed in TOPdesk.
+10. Optional: Check the *Close Mirrored TOPdesk Incident* integration parameter to close the TOPdesk incident when the corresponding Cortex XSOAR incident is closed.
```
```suggestion
9. Optional: Check the *Close Mirrored XSOAR Incident* integration parameter to close the Cortex XSOAR incident when the corresponding ticket is closed in TOPdesk.
10. Optional: Check the *Close Mirrored TOPdesk Ticket* integration parameter to close the TOPdesk ticket when the corresponding Cortex XSOAR incident is closed.
```

atossz replied:
changed

atossz replied:
done

BEAdi commented on Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.yml:
```diff
@@ -5827,6 +5827,45 @@ script:
+      secret: false
+    - default: false
+      defaultValue: '0'
+      description: The UTC timestamp in seconds of the last update. The incident or detection is only updated if it was
+        modified after the last update time.
```
```suggestion
      description: The UTC timestamp in seconds of the last update. The incident is only updated if it was
        modified after the last update time.
```

BEAdi commented on Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.yml:
```diff
@@ -5827,6 +5827,45 @@ script:
     outputs: []
+  - arguments:
+    - default: false
+      description: The remote incident or detection ID.
```
```suggestion
      description: The remote ticket ID.
```
Is it called a ticket or an incident? Ensure that everything has the same name.

atossz replied:
changed

atossz replied:
done

BEAdi commented on Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.yml:
```diff
@@ -5827,6 +5827,45 @@ script:
+      required: false
+      secret: false
+    deprecated: false
+    description: Gets remote data from a remote incident or detection. This method does not update
+     the current incident or detection, and should be used for debugging purposes only.
```
```suggestion
    description: Gets remote data from a remote incident. This method does not update
     the current incident, and should be used for debugging purposes only.
```

atossz replied:
done

BEAdi commented on Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.yml:
```diff
@@ -5827,6 +5827,45 @@ script:
+    name: get-remote-data
+  - arguments:
+    - default: false
+      description: Date string representing the local time. The incident or detection is only returned
+        if it was modified after the last update time.
```
```suggestion
      description: Date string representing the local time. The incident is only returned
        if it was modified after the last update time.
```

atossz replied:
changed

atossz replied:
done

BEAdi commented on Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.yml:
```diff
@@ -5827,6 +5827,45 @@ script:
+      required: false
+      secret: false
+    deprecated: false
+    description: Gets the list of incidents and detections that were modified since the last update time.
```
```suggestion
    description: Gets the list of incidents that were modified since the last update time.
```

atossz replied:
changed

atossz replied:
done

BEAdi commented on Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.yml:
```diff
@@ -5827,6 +5827,45 @@ script:
+    execution: false
+    name: get-modified-remote-data
+  - deprecated: false
+    description: Updates the remote incident or detection with local incident or detection changes. This method
+      is only used for debugging purposes and will not update the current incident or detection.
```
```suggestion
    description: Updates the remote incident with local incident changes. This method
      is only used for debugging purposes and will not update the current incident.
```

atossz replied:
changed

atossz replied:
done

BEAdi reviewed (changes requested):
Hi,
You made some great changes, thank you very much.

One question- are all the needed fields mapped in the existing incoming mapper?

atossz commented on Packs/TOPdesk/Integrations/TOPdesk/README.md:
```diff
@@ -51,7 +51,32 @@ Make sure you use the right account for your needs and that the account used has
+4. Optional: You can go to the *Incidents fetch query* or *Detections fetch query* parameter and select the query to fetch the incidents from TOPdesk.
+5. In the *Mirroring Direction* integration parameter, select in which direction the incidents should be mirrored:
+    - Incoming - Any changes in TOPdesk incidents (`state`, `status`, `tactics`, `techniques`, `objectives`, `tags`, `hosts.hostname`) will be reflected in XSOAR incidents.
+    - Outgoing - Any changes in XSOAR incidents will be reflected in TOPdesk incidents (`tags`, `status`).
```
changed

atossz replied:
done

atossz commented on Packs/TOPdesk/Integrations/TOPdesk/README.md:
```diff
@@ -51,7 +51,32 @@ Make sure you use the right account for your needs and that the account used has
+1. Navigate to __Settings__ > __Integrations__ > __Servers & Services__.
+2. Search for **TOPdesk** and select your integration instance.
+3. Enable **Fetches incidents**.
+4. Optional: You can go to the *Incidents fetch query* or *Detections fetch query* parameter and select the query to fetch the incidents from TOPdesk.
```
removed

atossz replied:
done

atossz commented on Packs/TOPdesk/Classifiers/classifier-TOPdesk-Incident.json:
```diff
@@ -0,0 +1,18 @@
+    "description": "",
+    "feed": false,
+    "id": "1097be01-8cc7-4594-8c63-ba92eca5dc8e",
+    "keyTypeMap": {},
```
changed

atossz replied:
done

atossz commented:
> Hi, You made some great changes, thank you very much.
> 
> One question- are all the needed fields mapped in the existing incoming mapper?

Yes, all available fields are now mapped from the incoming mapper.
I've change all the other documentation requests.

BEAdi reviewed (changes requested):
Thanks for making the changes so quickly! Waiting for the demo next week.

atossz commented on Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.yml:
```diff
@@ -5850,12 +5822,59 @@ script:
+    outputs: []
+  - arguments:
+    - default: false
+      description: The remote incident  ID.
```
done

atossz commented on Packs/TOPdesk/Integrations/TOPdesk/README.md:
```diff
@@ -59,17 +59,14 @@ To setup the mirroring follow these instructions:
-10. Optional: Check the *Close Mirrored TOPdesk Incident* integration parameter to close the TOPdesk incident when the corresponding Cortex XSOAR incident is closed.
+5. The *Comment Entry Tag*, *Work Note Entry Tag* and *File Entry Tag* integration parameters can be used to specify which comments and attachments should be mirrored to TOPdesk. When the tag *Comment Entry Tag* is used, the comment is visible to the operator and the person. If the tag *Work Note Entry Tag* is used, the comment is only visible to the operator and the tag *File Entry Tag* is used  ...
+6. Optional: Check the *Close Mirrored XSOAR Incident* integration parameter to close the Cortex XSOAR incident when the corresponding ticket is closed in TOPdesk.
+7. Optional: Check the *Close Mirrored TOPdesk Incident* integration parameter to close the TOPdesk ticket when the corresponding Cortex XSOAR incident is closed.
```
done

bakatzir commented on Packs/TOPdesk/Integrations/TOPdesk/TOPdesk_test.py:
```diff
@@ -929,10 +936,13 @@ def test_fetch_incidents(client, requests_mock, topdesk_incidents_override, last
         response_incident['number'] = incident_override['number']
         response_incident['creationDate'] = incident_override['creationDate']
+        response_incident['mirror_direction'] = None
+        response_incident['mirror_tags'] = ["comments", "ForServiceNow", "work_notes"]
```
```suggestion
        response_incident['mirror_tags'] = ["comments", "ForTopDesk", "work_notes"]
```

bakatzir commented on Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py:
```diff
@@ -1255,10 +1269,17 @@ def fetch_incidents(client: Client,
+        topdesk_incident['mirror_direction'] = MIRROR_DIRECTION.get(str(demisto_params.get('mirror_direction')))
+        topdesk_incident['mirror_tags'] = [
+            demisto_params.get('comment_tag', 'comments'),
+            demisto_params.get('file_tag', 'ForServiceNow'),
```
```suggestion
            demisto_params.get('file_tag', 'ForTopDesk'),
```

atossz commented on Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.yml:
```diff
@@ -38,6 +34,53 @@ configuration:
+  additionalinfo: Choose the tag to add to an entry to mirror it as a hidden comment in TOPdesk for the persons.
+  hidden: false
+- additionalinfo: Choose the tag to add to an entry to mirror it as a file in TOPdesk.
+  defaultvalue: ForServiceNow
```
done

atossz commented:
@BEAdi 
I've changed the layout as discussed with demo last week. I also changed some of the labels, which referred to ServiceNow and not TOPdesk.

The last item is the Entries from the beginning of the fetch Incident. I tried to look into it, but I wasn't able to resolve it. In the fetch_incidents, the incident is not created yet, so I couldn't add the entries. I tried to fetch the entries on the initial get_remote_data_command, but I wasn't able to fetch the created entries in TOPdesk before the fetch incident from Xsoar. Maybe you have some clues/examples how we could resolve it? I think something with the update timestamp, but which one...

BEAdi commented:
> I've changed the layout as discussed with demo last week. I also changed some of the labels, which referred to ServiceNow and not TOPdesk.

The changes you did are great!
Please mind the validation error that is now raised.

> The last item is the Entries from the beginning of the fetch Incident. I tried to look into it, but I wasn't able to resolve it. In the fetch_incidents, the incident is not created yet, so I couldn't add the entries. I tried to fetch the entries on the initial get_remote_data_command, but I wasn't able to fetch the created entries in TOPdesk before the fetch incident from Xsoar. Maybe you have some clues/examples how we could resolve it? I think something with the update timestamp, but which one...

On the initial `get_remote_data_command` it is indeed related to timestamps - [here](https://github.com/demisto/content/pull/18735/files#diff-04afe05571efac231716f3c3d6671b3e424eda86f67a9a0d9604ba4928628312R1346) you drop all the entries that were changed before the last update. It could be added there.
I think it will be a better solution to add it while fetching incidents - you can see [this example](https://github.com/demisto/content/blob/master/Packs/Jira/Integrations/JiraV2/JiraV2.py#L512), with the `get_entries_for_fetched_incident` function and the use of it (adding comments in `labels` to the incident, and attachments in `attachment`, which can be done [here](https://github.com/demisto/content/pull/18735/files#diff-04afe05571efac231716f3c3d6671b3e424eda86f67a9a0d9604ba4928628312R1281)).

BEAdi commented on Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py:
```diff
@@ -1277,10 +1277,26 @@ def fetch_incidents(client: Client,
                 'name': f"{topdesk_incident['briefDescription']}",
-                'details': json.dumps(topdesk_incident),
+                'labels': labels,
+                'details': f"{topdesk_incident['request']}",
```
Why was this changed?

atossz replied:
While I was comparing the code with JiraV2.py in fetch_incidents and create_incidents, I saw the incident there also puts the 'demisto.get(issue, "fields.description")' in the details field. It makes more sense, because before the complete json data was put there, while it was also stored in the rawJSON fields. It sounds like the same data on two places and no good description for the incident.

I see ServiceNow is dumping the complete JSON there also. So it is not consistent within all the fetch_incidents.

But the description sounds more logical and shows better in the layout.

BEAdi commented on Packs/TOPdesk/Integrations/TOPdesk/TOPdesk_test.py:
```diff
@@ -943,7 +948,7 @@ def test_fetch_incidents(client, requests_mock, topdesk_incidents_override, last
             expected_incidents.append({
                 'name': f"{incident_override['briefDescription']}",
-                'details': json.dumps(response_incident),
+                'details': f"{response_incident['request']}",
```
Why was this changed?

zumitnl replied:
While I was comparing the code with JiraV2.py in fetch_incidents and create_incidents, I saw the incident there also puts the 'demisto.get(issue, "fields.description")' in the details field. It makes more sense, because before the complete json data was put there, while it was also stored in the rawJSON fields. It sounds like the same data on two places and no good description for the incident.

I see ServiceNow is dumping the complete JSON there also. So it is not consistent within all the fetch_incidents.

But the description sounds more logical and shows better in the layout.
